airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java

```diff
@@ -455,17 +462,91 @@ public List<AutoCloseableIterator<AirbyteMessage>> getIncrementalIterators(final
-      return handler.getIncrementalIterators(catalog, tableNameToTable, emittedAt);
+      final List<AirbyteStateMessage> rawStateMessages = stateManager.getRawStateMessages();
+
+      final List<AirbyteStateMessage> statesFromCtidSync = new ArrayList<>();
+      final List<AirbyteStateMessage> statesFromXminSync = new ArrayList<>();
+
+      final Set<AirbyteStreamNameNamespacePair> alreadySeenStreams = new HashSet<>();
+      final Set<AirbyteStreamNameNamespacePair> streamsStillInCtidSync = new HashSet<>();
+
+      if (rawStateMessages != null) {
+        rawStateMessages.forEach(s -> {
+          final JsonNode streamState = s.getStream().getStreamState();
+          final StreamDescriptor streamDescriptor = s.getStream().getStreamDescriptor();
+          if (streamState == null || streamDescriptor == null) {
+            return;
+          }
+          final AirbyteStateMessage clonedState = Jsons.clone(s);
+          if (streamState.has("type") && streamState.get("type").asText().equalsIgnoreCase("ctid")) {
+            statesFromCtidSync.add(clonedState);
+            streamsStillInCtidSync.add(new AirbyteStreamNameNamespacePair(streamDescriptor.getName(), streamDescriptor.getNamespace()));
+          } else {
+            statesFromXminSync.add(clonedState);
+          }
+          alreadySeenStreams.add(new AirbyteStreamNameNamespacePair(streamDescriptor.getName(), streamDescriptor.getNamespace()));
+        });
+      }
+
+      final List<ConfiguredAirbyteStream> newlyAddedStreams = identifyNewlyAddedStreams(catalog, alreadySeenStreams);
+      final List<ConfiguredAirbyteStream> streamsForCtidSync = new ArrayList<>();
+      catalog.getStreams().stream()
+          .filter(stream -> streamsStillInCtidSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
+          .map(Jsons::clone)
+          .forEach(streamsForCtidSync::add);
+
+      streamsForCtidSync.addAll(newlyAddedStreams);
+
+      final List<ConfiguredAirbyteStream> streamsForXminSync = catalog.getStreams().stream()
+          .filter(stream -> !streamsForCtidSync.contains(stream))
+          .map(Jsons::clone)
+          .toList();
+
+      LOGGER.info("Streams to be synced via ctid : {}", streamsForCtidSync.size());
+      LOGGER.info("Streams to be synced via xmin : {}", streamsForXminSync.size());
+
+      final List<AutoCloseableIterator<AirbyteMessage>> ctidIterator = new ArrayList<>();
+      final List<AutoCloseableIterator<AirbyteMessage>> xminIterator = new ArrayList<>();
+
+      if (!streamsForCtidSync.isEmpty()) {
+        final CtidStateManager ctidStateManager = new CtidStateManager(statesFromCtidSync);
+        final PostgresCtidHandler ctidHandler = new PostgresCtidHandler(database, sourceOperations, getQuoteString(), ctidStateManager,
+            x -> new ObjectMapper().valueToTree(xminStatus),
```

Comment:
nit : IMO it's super confusing to have lambdas as function arguments here and above. can we calculate them seperately and pass them in?